flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java

```diff
@@ -254,6 +262,58 @@ public void failJobIfNotEnoughResources() throws Exception {
 				"Could not allocate the required slot within slot request timeout.").isPresent());
 	}
 
+	@Test
+	public void vertexIsResetBeforeRestarted() throws Exception {
+		final JobGraph jobGraph = singleNonParallelJobVertexJobGraph();
+
+		final TestSchedulingStrategy.Factory schedulingStrategyFactory = new TestSchedulingStrategy.Factory();
+		final DefaultScheduler scheduler = createScheduler(jobGraph, schedulingStrategyFactory);
+		final TestSchedulingStrategy schedulingStrategy = schedulingStrategyFactory.getLastCreatedSchedulingStrategy();
+		final SchedulingTopology topology = schedulingStrategy.getSchedulingTopology();
+
+		startScheduling(scheduler);
+
+		final SchedulingExecutionVertex vertex = Iterables.getOnlyElement(topology.getVertices());
+		final ExecutionAttemptID attemptId = scheduler.getExecutionVertex(vertex.getId()).getCurrentExecutionAttempt().getAttemptId();
+
+		schedulingStrategy.schedule(Collections.singleton(vertex.getId()));
+
+		scheduler.updateTaskExecutionState(new TaskExecutionState(jobGraph.getJobID(), attemptId, ExecutionState.FAILED));
+
+		taskRestartExecutor.triggerScheduledTasks();
+
+		Set<ExecutionVertexID> verticesToRestart = schedulingStrategy.getReceivedVerticesToRestart();
+		assertEquals(1, verticesToRestart.size());
+		assertEquals(ExecutionState.CREATED, topology.getVertexOrThrow(verticesToRestart.iterator().next()).getState());
+	}
+
+	@Test
+	public void scheduleOnlyIfVertexIsCreated() throws Exception {
+		final JobGraph jobGraph = singleNonParallelJobVertexJobGraph();
+
+		final TestSchedulingStrategy.Factory schedulingStrategyFactory = new TestSchedulingStrategy.Factory();
+		final DefaultScheduler scheduler = createScheduler(jobGraph, schedulingStrategyFactory);
+		final TestSchedulingStrategy schedulingStrategy = schedulingStrategyFactory.getLastCreatedSchedulingStrategy();
+		final SchedulingTopology topology = schedulingStrategy.getSchedulingTopology();
+
+		startScheduling(scheduler);
+
+		final SchedulingExecutionVertex vertex = Iterables.getOnlyElement(topology.getVertices());
+
+		// The first time scheduling of a vertex will result in one deployment
+		schedulingStrategy.schedule(Collections.singleton(vertex.getId()));
+		testTaskManagerGateway.getDeployedExecutionVertices(1, TIMEOUT_MS);
+
+		// The scheduling of a non-CREATED vertex will result in no deployment
+		schedulingStrategy.schedule(Collections.singleton(vertex.getId()));
+		try {
+			testTaskManagerGateway.getDeployedExecutionVertices(1, TIMEOUT_MS);
```

Comment:
It's probably better to use the changes from #9872 later

Reply:
Yes. I like that change in #9872.

Reply:
But maybe that depends on which PR can be merged first. I think doing the change in any PR should be easy.